tests/discrepancy/test_cd.py

```diff
@@ -0,0 +1,201 @@
+        ValueError, match="Both propensity model and propensity estimates are specified"
+    ):
+        cd_func.condind(
+            X=X,
+            Y=Y,
+            group_ind=group_ind,
+            propensity_model=RandomForestClassifier(),
+            propensity_weights=[0.5, 0.5],
+            random_seed=seed,
+        )
+
+    Y = sample_df[y].to_numpy()
+    X = sample_df[x].to_numpy()
+    with pytest.raises(ValueError, match="There are 3 group pre-defined estimates"):
+        cd_func.condind(
+            X=X,
+            Y=Y,
+            group_ind=group_ind,
+            propensity_weights=np.ones((10, 3)) * 0.5,
+            random_seed=seed,
+        )
+
+    with pytest.raises(ValueError, match="There are 100 pre-defined estimates"):
+        cd_func.condind(
+            X=X,
+            Y=Y,
+            group_ind=group_ind,
+            propensity_weights=np.ones((100, 2)) * 0.5,
+            random_seed=seed,
+        )
+
+
+@pytest.mark.parametrize(
+    ["cd_func", "cd_kwargs"],
+    [
+        [bregman, dict()],
+        [kcd, dict()],
+        [bregman, {"propensity_model": RandomForestClassifier(n_estimators=50, random_state=seed)}],
+        [bregman, {"propensity_weights": np.ones((n_samples, 2)) * 0.5}],
+        [kcd, {"propensity_model": RandomForestClassifier(n_estimators=50, random_state=seed)}],
+        [kcd, {"propensity_weights": np.ones((n_samples, 2)) * 0.5}],
+        # [kcd, {"l2": 1e-3}],
+        # [kcd, {"l2": (1e-3, 2e-3)}],
+    ],
+)
+@pytest.mark.parametrize(
+    ["df", "env_type"],
+    [
+        [single_env_scm(n_samples=n_samples, offset=2.0), "single"],
+        [multi_env_scm(n_samples=n_samples // 2, offset=2.0), "multi"],
+    ],
+)
+def test_cd_simulation(cd_func, df, env_type, cd_kwargs):
+    """Test conditional discrepancy tests."""
+    group_col = "group"
+    alpha = 0.1
+    null_sample_size = 25
+
+    if env_type == "single":
+        res = cd_func.condind(
+            X=df["x"],
+            Y=df["x1"],
+            group_ind=df[group_col],
+            null_sample_size=null_sample_size,
+            **cd_kwargs,
+            random_seed=seed,
+        )
+        assert res.pvalue > alpha, f"Fails with {res.pvalue} not greater than {alpha}"
+        res = cd_func.condind(
+            X=df["x"],
+            Y=df["z"],
+            group_ind=df[group_col],
+            null_sample_size=null_sample_size,
+            n_jobs=-1,
+            **cd_kwargs,
+            random_seed=seed,
+        )
+        assert res.pvalue > alpha, f"Fails with {res.pvalue} not greater than {alpha}"
+        res = cd_func.condind(
+            X=df["x"],
+            Y=df["y"],
+            group_ind=df[group_col],
+            null_sample_size=null_sample_size,
+            n_jobs=-1,
+            **cd_kwargs,
+            random_seed=seed,
+        )
+        assert res.pvalue > alpha, f"Fails with {res.pvalue} not greater than {alpha}"
+    elif env_type == "multi":
```

Comment:
What about instead of the if-elifs, just separate them into two tests with `test_cd_simulation_single` and `test_cd_simulation_multi`.

Optimally, one would even have a separate test for each `assert` (but we don't need to over-engineer it).

Also, since you don't use the given-when-then pattern for the unit test names, add a brief description what is tested here.

Reply:
Refactored and also renamed tests.